game-app/game-core/src/main/java/games/strategy/triplea/ai/pro/ProCombatMoveAi.java

```diff
@@ -1202,6 +1162,86 @@ && canAirSafelyLandAfterAttack(unit, t)) {
     }
   }
 
+  private void determineTerritoriesThatCanBeBombed(
+      final Map<Territory, ProTerritory> attackMap,
+      final Map<Unit, Set<Territory>> sortedUnitAttackOptions,
+      final Set<Unit> alreadyAttackedWithUnits) {
+    final Map<Unit, Set<Territory>> bomberMoveMap =
+        territoryManager.getAttackOptions().getBomberMoveMap();
+    final Set<Unit> bombers = bomberMoveMap.keySet();
+    bombers.removeAll(alreadyAttackedWithUnits);
```

Comment:
I think this changes the underlying map. Is this really intended?

Honestly, I find the previous formulation more straight forward. It seems OK to just skip keys that are in the `alreadyAttackedWithUnits` - and the performance of that is easier to reason about than removeAll(), especially if it would require us to make a copy.

Plus, if you iterate over entries, you can avoid the bomberMoveMap.get() lookup you have on line 1176.

Reply:
Whole section reworked also taking into account the other comment.